### PR TITLE
Blaze: Fix state change warnings when editing ads

### LINKED ISSUE #11780: Blaze: Hitting character limit causes warning for "Publishing changes from within view updates"
image

image

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #11780 

## Description

In SwiftUI, if we try to change a view's state just before it's about to change, we get the warning "Publishing changes from within view updates is not allowed".

In `BlazeEditAdViewModel`, we violated this in two places:
- We observe `tagline` to change `taglineRemainingLength` immediately. Please note, we get a signal for change of `tagline` just before the value is changed. When we change `taglineRemainingLength` in `sink`, we are triggering an update to the SwiftUI view that listens to the property, before the the view even updates following the change to `tagline`.
- On `BlazeEditAdView`, we observe `tagline` with `onChange` and trigger the update to the value if the text exceeds the character limit. This again triggers state change even before the view is updated for `tagline`.

Solution:
- Updated `taglineRemainingLength` and `taglineEmptyError` using `map` instead of `sink` when observing `tagline`. 
- Added a similar fix for `descriptionRemainingLength` and `descriptionEmptyError`.
- Removed the forced formatting of tagline and description. Instead, we should let the user continue typing, while displaying the character count in red to signal an error.

## Testing instructions

- Build the app and log in to a store eligible for Blaze.
- Select Promote in the Blaze secion on the Dashboard screen.
- Select Edit Ad to change the content.
- Type in the tagline and description fields until you exceed the character limits. Confirm the following:
  - The text should be displayed as is
  - The character count should be red when the text length exceeds the limit.
  - No warnings on Xcode regarding `Publishing changes from within view updates is not allowed`.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/5533851/071cb827-f5b7-49d5-ae18-32c664367dde

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/182">17.1</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdViewModel.swift [hunk 1]
```diff
@@ -21,33 +21,37 @@ final class BlazeEditAdViewModel: ObservableObject {
...
     }
-    private var descriptionEmptyError: String?
+
+    @Published private var descriptionEmptyError: String?
 
     var isSaveButtonEnabled: Bool {
-        guard let editedAdData else {
+        guard let editedAdData,
+                taglineRemainingLength >= 0,
+                descriptionRemainingLength >= 0 else {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdViewModel.swift:69]: nit: Kinldly consider adding unit tests for this.

File: WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdView.swift [hunk 2]
```diff
@@ -118,6 +115,7 @@ private extension BlazeEditAdView {
             )
 
             Text(viewModel.taglineFooterText)
+                .foregroundStyle(viewModel.taglineRemainingLength >= 0 ? .secondary : Color(.error))
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdView.swift]: Nice work using the `error` color. 

Should we use `.error` color also when the tagline is empty? Applies to the description field as well.

**Review by selanthiraiyan — APPROVED** [review]
Thanks for the fix and for explaining in detail the reason and solution. 🙇

File: WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdViewModel.swift (hunk 1 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdViewModel.swift:69]: Added tests in 444424c9cdfe5eefe3b876072ff89801007207ee.

File: WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdView.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdView.swift]: Good point - I added properties for validating the fields and updated the view to use them in 34aaf628f909d6b2561f8adcb6c78e7792b867b1. Test added in 053d536e370b8bd1184126a70b7407f7298beb1b.

**Comment by itsmeichigo** [comment]
Thank you, @selanthiraiyan, for the reviews! I went ahead and addressed the issues you mentioned in the comment in the subsequent PRs. Also, I realized that the messages for remaining characters with a minus number are not user friendly, so I also added a change to use a generic error message in 102dbb3e756514dae8ad43fe3c69bdf8bd8a4cc3 (with tests in 9e5214bd5e06d39d4340de7318b56f40b3728055). The messages should now be like this:



I'll enable auto-merge for now - please feel free to leave any comments and I'll address them in another PR if needed.

**Pull request merged**